runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp

```diff
@@ -0,0 +1,1536 @@
+
+#include "j9.h"
+#include "jclprots.h"
+#include "j9cp.h"
+#include "j9protos.h"
+#include "ut_j9jcl.h"
+#include "j9port.h"
+#include "jclglob.h"
+#include "jcl_internal.h"
+#include "util_api.h"
+#include "j9vmconstantpool.h"
+#include "ObjectAccessBarrierAPI.hpp"
+#include "objhelp.h"
+
+#include <string.h>
+#include <assert.h>
+
+#include "VMHelpers.hpp"
+
+extern "C" {
+
+/* Constants mapped from java.lang.invoke.MethodHandleNatives$Constants */
+#define MN_IS_METHOD		0x00010000
+#define MN_IS_CONSTRUCTOR	0x00020000
+#define MN_IS_FIELD			0x00040000
+#define MN_IS_TYPE			0x00080000
+#define MN_CALLER_SENSITIVE	0x00100000
+
+#define MN_REFERENCE_KIND_SHIFT	24
+#define MN_REFERENCE_KIND_MASK	0xF
+
+#define MN_SEARCH_SUPERCLASSES	0x00100000
+#define MN_SEARCH_INTERFACES	0x00200000
+
+/* Private MemberName object init helper 
+ *
+ * Set the MemberName fields based on the refObject given:
+ * For j.l.reflect.Field:
+ *		find JNIFieldID for refObject, create j.l.String for name and signature and store in MN.name/type fields.
+ *		set vmindex to the fieldID pointer and target to the J9ROMFieldShape struct.
+ *		set MN.clazz to declaring class in the fieldID struct.
+ * For j.l.reflect.Method or j.l.reflect.Constructor:
+ *		find JNIMethodID, set vmindex to the methodID pointer and target to the J9Method struct.
+ *		set MN.clazz to the refObject's declaring class.
+ *
+ * Then for both, compute the MN.flags using access flags and invocation type based on the JNI-id.
+ *
+ * Throw an IllegalArgumentException if the refObject is not a Field/Method/Constructor
+ */
+void
+initImpl(J9VMThread *currentThread, j9object_t membernameObject, j9object_t refObject)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+
+	J9Class* refClass = J9OBJECT_CLAZZ(currentThread, refObject);
+
+	jint flags = 0;
+	jlong vmindex = 0;
+	jlong target = 0;
+	j9object_t clazzObject = NULL;
+	j9object_t nameObject = NULL;
+	j9object_t typeObject = NULL;
+
+	if (refClass == J9VMJAVALANGREFLECTFIELD(vm)) {
+		J9JNIFieldID *fieldID = vm->reflectFunctions.idFromFieldObject(currentThread, NULL, refObject);
+		vmindex = (jlong)fieldID;
+		target = (jlong)fieldID->field;
+
+		flags = fieldID->field->modifiers & CFR_FIELD_ACCESS_MASK;
+		flags |= MN_IS_FIELD;
+		flags |= (J9_ARE_ANY_BITS_SET(flags, J9AccStatic) ? MH_REF_GETSTATIC : MH_REF_GETFIELD) << MN_REFERENCE_KIND_SHIFT;
+		J9UTF8 *name = J9ROMFIELDSHAPE_NAME(fieldID->field);
+		J9UTF8 *signature = J9ROMFIELDSHAPE_SIGNATURE(fieldID->field);
+
+		nameObject = vm->memoryManagerFunctions->j9gc_createJavaLangString(currentThread, J9UTF8_DATA(name), (U_32)J9UTF8_LENGTH(name), J9_STR_INTERN);
+
+		PUSH_OBJECT_IN_SPECIAL_FRAME(currentThread, nameObject);
+		typeObject = vm->memoryManagerFunctions->j9gc_createJavaLangString(currentThread, J9UTF8_DATA(signature), (U_32)J9UTF8_LENGTH(signature), J9_STR_INTERN);
```

Comment:
Does `typeObject` need to be a string or can you set the class directly?  ie:
```suggestion
		typeObject =J9VMJAVALANGREFLECTFIELD_TYPE(vmThread, refObject);
```